src/Agent.Worker/StepsRunner.cs

```diff
@@ -205,5 +204,36 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 Trace.Info($"Current state: job state = '{jobContext.Result}', step failed = {stepFailed}, critical step failed = {criticalStepFailed}");
             }
         }
+
+        private bool ShouldRunStep(IStep step, bool stepFailed, bool criticalStepFailed)
+        {            
+            switch(step.RunMode)
+            {
+                // The server hasn't been upgraded yet. Fallback to old code
+                case RunMode.Undefined:
```

Comment:
...or rather, can the legacy properties be deleted from istep and taskrunner handle creating an appropriate RunMode based on the TaskInstance properties. would that work?